### DISCUSSION TITLE
Incorrect behavior with URL query params

### DISCUSSION THREAD

**Discussion by xswebfr** [livewire/livewire#8791] [7. Bugs (archived)]
### Livewire version

v3.5.9 (previously with v3.5.6)

### Laravel version

v11.26 (previously with v10.48.20)

### Which PHP version are you using?

PHP 8.3 (previously with PHP 8.2)

### Steps To Reproduce

When using #[Url] (or queryString()) on properties, the query string has a different behavior depending on whether the property is in the URL on page loading or not.

Exemple #1 :
Navigate to https://4jn31.wirebox.app:8443/ in wirebox
Click on button 5 : https://4jn31.wirebox.app:8443/?param=5
Click on button null : https://4jn31.wirebox.app:8443/
Click on button 5 : https://4jn31.wirebox.app:8443/?param=5
Click on button null : https://4jn31.wirebox.app:8443/
=> It works as expected

Exemple #2 :
Navigate to https://4jn31.wirebox.app:8443/?param=10 in wirebox
Click on button null : https://4jn31.wirebox.app:8443/?param=
Click on button 5 : https://4jn31.wirebox.app:8443/?param=5
Click on button null : https://4jn31.wirebox.app:8443/?param=
=> ?param= always stays

I tried to play with Url properties (keep, except etc) but it has no effect.

### Repository or Wirebox URL

https://wirebox.app/b/4jn31

### Additional information

Full examples in wirebox

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by xswebfr** [comment]
This issue has been occurring since version 3.0, I think.

**Comment by aoxborrow** [comment]
yeah this has been driving me nuts today, if you load the page with query string properties, you can't clear them from the query string if you use nullable type or property without a default value, e.g.

```php
public ?int $a = null; // ?a= will remain
public $b; // ?b= will remain
```

the only way to clear from the query string is if you use an actual default value, and a matching `except`:
```php
public int $a = 0;
public $b = '';

protected function queryString() {
    return [
        'a' => ['except' => 0],
        'b' => ['except' => ''],
    ];
}

// or with attribute works the same way
#[Url(except: 0)]
public int $a = 0;

#[Url(except: '')]
public $b = '';
```

I have 20+ filters so this is all very verbose, not only do i have to put a default value (instead of nullable), but I have to put it again in `except`.

looks like @joshhanley was trying to fix this in #7779, but it got reverted

**Reply by aoxborrow** [reply]
also, loading the page with any blank query string values triggers a PHP Warning:
```
json_decode(): Passing null to parameter #1 ($json) of type string is deprecated in 
../vendor/livewire/livewire/src/Features/SupportQueryString/BaseUrl.php on line 65
```

**Comment by xswebfr** [comment]
I updated my stack to PHP 8.3, Laravel 11.26 and livewire 3.5.9 and it still happens FYI.

**Comment by bayukartiko** [comment]
any updates?

**Comment by bayukartiko** [comment]
> Exemple #2 :
Navigate to https://4jn31.wirebox.app:8443/?param=10 in wirebox
Click on button null : https://4jn31.wirebox.app:8443/?param=
Click on button 5 : https://4jn31.wirebox.app:8443/?param=5
Click on button null : https://4jn31.wirebox.app:8443/?param=
=> ?param= always stays

I’ve implemented a temporary workaround for that issue.

`app.blade.php`
```html
    <!-- bottom of the HTML page -->
        <script type="text/javascript">
            window.addEventListener('removeQueryString', parameter => {
                $(document).ready(function(){
                    var url = document.location.href;
                    var urlparts = url.split('?');
                    if (urlparts.length >= 2) {
                        var urlBase = urlparts.shift();
                        var queryString = urlparts.join("?");
                        var pars = queryString.split(/[&;]/g);
                        parameter.detail?.forEach(parameter => {
                            var prefix = encodeURIComponent(parameter) + '=';
                            for (var i = pars.length; i-- > 0;) {
                                if (pars[i].lastIndexOf(prefix, 0) !== -1) {
                                    pars.splice(i, 1);
                                }
                            }
                        });
                        if (pars.length === 0) {
                            url = urlBase;
                        } else {
                            url = urlBase + '?' + pars.join('&');
                        }
                        window.history.replaceState('', document.title, url);
                    }
                    return url;
                });
            });
        </script>
    </body>
</html>
```

Inside your Livewire component, you can dispatch the event like this:
`$this->dispatch('removeQueryString', 'param', 'otherParam', ...);`

This may not be the optimal solution, but it serves as a temporary fix while awaiting the PR for a better solution that doesn't require manual hardcoding.
